### ISSUE TITLE
Little random glitch that holds the PWM pulse high (100% duty) from time to time

### ISSUE THREAD

**Issue by rhormigo** [MajicDesigns/MD_PWM#1]
--Tried in last version September 1.0.3
--No found other reports on this issue or any others (but my use of github is limited), I may be missing if there is a discussion somewhere.

There is a little random glitch that holds the PWM pulse high (100% duty) from time to time (for just a pulse appears), it was hard to catch in the beginning, but using a second instance of the library as trigger clock, I can see it easily with a digital oscilloscope setting the pwm frequency from 50 to 300Hz, either with timer 1 or 2, and nothing else running. It only happens when sequencing duty cycle from high to low, with a for loop from write(255) to write(0) for example. Going up, or holding constant is ok.
  
Environment is running IDE 1.8.13.over Mega on Windows10 with a very ok computer, no issues in compilation, etc... 
I am currently making a quick fork for some school research that convert the class in a PWM DC motor driver for PWM modules that use enable, PWM, and direction. and I noticed this when decreasing the motor speed (pwm decreasing in steps of 50 even 200ms is visible). So I went back to your basic example and place this for loop there, and yes, there still is it.

This is my for loop below, if you go increasing instead, then is fine! Faster loop also makes the issue more visible:
      for (uint8_t i=255; i>=0; i--){
      pwm[1].write(i);
      delay(200); //Reducing further than 200 makes things worst
    }
Complete code here:
[MD_PWM_Test.txt](https://github.com/MajicDesigns/MD_PWM/files/7232795/MD_PWM_Test.txt)

I also found a solution that appears to work, at least for a few minutes that I have been testing so far:
I saw that the timer ISR calls the  setPin function every 52uS at 75Hz, that even at that low 75Hz is really demanding for this little16MHz CPU. It appears in my limited research, that this line:
if (_cycleCount == _pwmDuty && _pwmDuty != 0xff)  misses (skips?) the _pwmDuty  value, and stays on.  So I just modified like this:
if (_cycleCount >= _pwmDuty && _pwmDuty != 0xff) and now works. 
This I made intuitively, without understanding exactly why this happens, the ISR doesn't appear disturbed enough (at least toggling a pin from it) to miss a call.   Just to test, I disabled irq's while in the ISR to test, and same result. 

Anyway my two cents so far, please let me know if I am missing some, or possibly breaking some else... I'll keep updating if I see some else.
 
Rick.

**Comment by MajicDesigns** [comment]
Here's what I think is happening: Changing the PWM frequency does not reset the pwmCount, so it is possible that when the new pwmDuty value was set by write() the pwmCount value was already greater than pwmDuty for the new frequency.  pwmDuty (uint8_t which has values 0- 255) then gets incremented until it cycles around again at 255 and gets back in sync, so you would see the value remain high for a full 255 count cycle and then work again. 

Changing the == to >= makes sure that this condition is reset as soon as it it detected.

Keep testing and let me know.

**Comment by rhormigo** [comment]
"Changing the PWM frequency does not reset the pwmCount"
I imagine you mean changing the pwmDuty (duty cycle), as the frequency is always constant in my tests, the file I attached at 50Hz. That is a rate of 20mS per pwm pulse. The little test I posted shows glitches even when decreasing  the pwmDuty at a rate  of 200mS that is 10 times slower, and so multiple pulses are run over each pwmDuty value to be detected with == regardless you are decreasing or increasing. So still a bit confusing to me...  Now, I think to understand, the ISR is run 256 times faster to keep the count, like 78.125uS at 50Hz. Then it has to run all instances in that interval (4 in the test sketch) , that let us less than 20uS to run each instance call, including their if conditions, and  the digitalWrites()?   Sounds busy, reeeeeally busy for that little AVR :-).... so I wonder if sometimes just can't handle it, even with multiple cycles around.  

Anyway, so far so good with the >= change, that basically forces the pin down to the end of the cycle, works. Not very elegant, as I imagine this will add even extra CPU cycles.... Perhaps a faster digitalWrite right on the register may help down the road?
Also as another observation,  the loop in the ISR that checks for instances null pointers to see if these pins exist, not sure if could be avoided by keeping track of the instances externally to the ISR so you save this operation that is done every 20uS? 
   
I'll keep an eye on this, as this will be part of a huge Frankenstein project that involves lots of resources on a Mega. I can see trouble down the road when the interrupt in this module start fighting with others. I also have servo, and custom tone modules, all using different timers and resources, but the same CPU!

**Comment by MajicDesigns** [comment]
Yes, I meant the duty cycle changing with write().

"Not very elegant, as I imagine this will add even extra CPU cycles"

Pretty much came to the same conclusion thinking about this overnight. The >= means that for every pass in the second part of the cycle it will do a digitalWrite(). A more efficient fix is to change the comparison back to == and change write() to
`    inline void write(uint8_t duty) { _pwmDuty = duty; if(_pwnDuty > _cycleCount) _cycleCount = _pwmDuty; }
`
This should ensure that the digitalWrite() only happens once on the transition.

You can read about how this works this library at [https://arduinoplusplus.wordpress.com/2021/04/06/pwm-on-any-i-o-pin/](url) if you have not already done so.

**Comment by MajicDesigns** [comment]
I set up a test system and check a few things. My initial diagnosis was correct - when the duty point is moved to before the current count (ie the current count becomes greater than the new duty transition point), it would miss the transition. However, the fix needs to be a bit more subtle than has been discussed above.

Also note that 'the glitch' only occurs when the new duty cycle is less than the current duty cycle.

There are 2 situations that can arise when the duty cycle is changed:
1. The current cycle count is already past the current duty transition point. In this case the digital transition has already happened and we don't have to do anything. 
2. The current count is still less than the new duty transition point. Digital transition is still top happen and we also don't have to do anything.
3. The current count is less than the old duty point but greater than the new duty point (ie it now sits in-between). In this case the transition should have happened but will be missed in setPin() - this is 'the glitch'. A simple way to handle this is to check for it in write() and reset the current count to the new duty count. This extends the duration of the current PWM cycle but provides the correct duration for the LOW portion.

The library has been changed to implement this strategy and it seems to pretty smoothly from what I can see on my 'scope. Uploaded this fix in v 1.0.4.

**Comment by rhormigo** [comment]
Hi, thanks for the quick responses on this... this is a good fixing to get a correction just in the spot without needing to go to the next cycle to fix it. But yet some happen in the setPin() that misses the == 
It looks less visible than yesterday, especially at slow rate of change, takes a while to show with a pause of 200mS between writes,  but yet I can see the glitch there when decrementing...  at faster sampling the glitch still comes often. See attached video based in the same example from yesterday. Check glitches at second 13 and at the very end:
  https://user-images.githubusercontent.com/35308546/135021774-88b9710e-f11e-4a59-9b82-f5f2ed381a5c.mp4

Rick.

**Comment by MajicDesigns** [comment]
Definitely not seeing the same as you do on my scope. The PWM duty cycle is really steady here.
1. My scope is definitely not as good as yours, so it may be missing what you see. However, I was seeing the original problem pretty much as per your video.
2. Are you sure that you have the latest source code?  How are you retrieving the code from github?
3. The only thing I can think of is some kind of boundary condition, Maybe try changing this line in write() as per below
`   if (_cycleCount <= _pwmDuty && // the current count has not yet caused the digital transition ..
`

**Comment by rhormigo** [comment]
1 and 2 yes. 
3- YESSS, that damn = condition was the one definitely missed at the boundary, that is why was was happening at a very low rate now, and your difficulty to see it, I imagine.  I'll keep an eye as I get more time, but a quick test with 2 and 3 pins show no glitches now. Rock stable!  I even test setting my delay in the for loop to really fast changes (from 1mS) and still works, before under 10mS pwm duty changes was getting bad. This is the example I am using:
 [MD_PWM_Test.txt](https://github.com/MajicDesigns/MD_PWM/files/7246559/MD_PWM_Test.txt)

Now, this doesn't bother me... but there is an interesting side effect that looks to keep track of the corrected glitches. See my new video: With the same example file, I leave one pin as stable timing reference to trigger the scope. The two other pins with a fast changing duty (every 5mS). I am speculating that every time was a miss and glitch before, the phase of the pulse goes one time slot behind (78uS in my example at 50Hz). This also happens at slower changing duty, but slower. The events where this happen are not consistent across instances, so a race of pulses is created :-) At 5ms delay is quickly visible[ (see video)](https://www.dropbox.com/s/lv867yvcrqvhbs7/race.mp4?dl=0) The pulse delay over the stable reference is quickly visible, but the race between them starts around minute 1, after 15 minutes pulses get totally out of phase (video is two first minutes). 

I'll keep playing with this at another time, as this is barely an issue, in fact I don't see at this moment this to be an issue in 99% of applications, including no for me and the DC motor driver I did. However, this opens old memories, as I actually had an issue with racing pwms in the past in an optogenetic platform I designed for a research project that used an huge array of power LEDs. I was using an array of 16 channel TI I2C drivers and there was a racing condition between drivers. It was creating a distracting funny aliasing with flickering effects. Fortunately these drivers had external clocking, and I just added a clock bus for all to be sync, and problem gone.  

Thanks for all the help.

**Comment by MajicDesigns** [comment]
The PWM was never guaranteed to be in sync across all the pins, and with the current changes it is even less likely that this is the case, as we are extending the 'low' portion of signal when we need to adjust.

I'll leave it up to you to close the issue.

**Comment by rhormigo** [comment]
Fair enough, then I was thinking a bit over dinner on this (can't help myself!) and came with a couple ideas to optimize the code around the ISR to free extra time and  then to feature synchronized pulses at the same time.
Instead to maintain a counter by instance, I just created a single global one gCnt, then the instances sync to time slot 0 when also they load the desired duty, then just compare with the == as before.
Also I saved around 5uS X #instances  at the ISR for each toggling  high or low of the pulse with direct register writing, what I hope helps some specially at higher pulse frequency or with other IRQs active.
I have attached code with my mods marked with //RH ...  Note that in my file I set the library includes to run in the same directory with the test sketch, just for this test purpose.
I tested in Mega  at different speeds from 1mS to 200mS by duty change, frequencies up to 300Hz, and even I pushed the code up to 5 pins, if there is no other irqs, I imagine can go well beyond that.. all good so far
So feel free to implement whatever you like and we are done, now I think I can go to sleep :-) (US EST here)
[MD_PWM_Test.zip](https://github.com/MajicDesigns/MD_PWM/files/7248610/MD_PWM_Test.zip)
